ImmuDB4Net/ImmuClientSync.cs

```diff
@@ -0,0 +1,1364 @@
+        : this(NewBuilder().WithServerUrl(serverUrl).WithServerPort(serverPort))
+    {
+    }
+
+    public ImmuClientSync(string serverUrl, int serverPort, string database)
+        : this(NewBuilder().WithServerUrl(serverUrl).WithServerPort(serverPort).WithDatabase(database))
+    {
+    }
+
+    internal ImmuClientSync(ImmuClientSyncBuilder builder)
+    {
+        ConnectionPool = builder.ConnectionPool;
+        GrpcAddress = builder.GrpcAddress;
+        connection = releasedConnection;
+        SessionManager = builder.SessionManager;
+        DeploymentInfoCheck = builder.DeploymentInfoCheck;
+        serverSigningKey = builder.ServerSigningKey;
+        stateHolder = builder.StateHolder;
+        heartbeatInterval = builder.HeartbeatInterval;
+        ConnectionShutdownTimeout = builder.ConnectionShutdownTimeout;
+        stateHolder.DeploymentInfoCheck = builder.DeploymentInfoCheck;
+        stateHolder.DeploymentKey = Utils.GenerateShortHash(GrpcAddress);
+        stateHolder.DeploymentLabel = GrpcAddress;
+    }
+
+
+    public static void ReleaseSdkResources()
+    {
+        RandomAssignConnectionPool.Instance.Shutdown();
+    }
+
+    private void StartHeartbeat()
+    {
+        heartbeatTask = Task.Factory.StartNew(async () =>
+        {
+            while (true)
+            {
+                if (heartbeatCloseRequested!.WaitOne(heartbeatInterval)) return;
+                try
+                {
+                    await Service.KeepAliveAsync(new Empty(), Service.GetHeaders(ActiveSession));
+                    heartbeatCalled?.Set();
+                }
+                catch (RpcException) { }
+            }
+        });
+    }
+
+    private void StopHeartbeat()
+    {
+        if (heartbeatTask == null)
+        {
+            return;
+        }
+        heartbeatCloseRequested?.Set();
+        heartbeatTask.Wait();
+        heartbeatCloseRequested?.Close();
+        heartbeatCalled?.Close();
+        heartbeatTask = null;
+    }
+
+    public void Open(string username, string password, string defaultdb)
+    {
+        try
+        {
+            using (ManualResetEvent mre = new ManualResetEvent(false))
+            {
+                while (true)
+                {
+                    if (Interlocked.Exchange(ref sessionSetupInProgress, 1) == 0)
+                    {
+                        break;
+                    }
+                    mre.WaitOne(2);
+                }
+            }
+            if (ActiveSession != null)
+            {
+                throw new InvalidOperationException("please close the existing session before opening a new one");
+            }
+            Connection = ConnectionPool.Acquire(new ConnectionParameters
+            {
+                Address = GrpcAddress,
+                ShutdownTimeout = ConnectionShutdownTimeout
+            });
+            ActiveSession = SessionManager.OpenSession(Connection, username, password, defaultdb);
```

Comment:
Looking at this code now, there's SessionManager.OpenSession call now. Maybe we could use the sync version in the Async immuclient too to also avoid sessionSetupInProgress lock and use a simple mutex? Are there any risks of mixing sync and async GRPC calls in same code? IIRC the sessionSetupInProgress lock was only added to async immuclient since we couldn't use a normal lock there.